src/HealthGPS/kevin_hall_model.cpp

```diff
@@ -195,53 +253,110 @@ void KevinHallModel::set_energy_intake(Person &person) const {
-    const double G_0 = person.get_risk_factor_value("Glycogen"_id);
-    const double EI_0 = person.get_risk_factor_value("EnergyIntake"_id);
-    const double CI_0 = person.get_risk_factor_value("Carbohydrate"_id);
+void KevinHallModel::initialise_kevin_hall_state(Person &person,
+                                                 std::optional<double> adjustment) const {
 
-    // TODO: Compute height.
-    double H = H_0;
+    // Apply optional weight adjustment.
+    if (adjustment.has_value()) {
+        person.risk_factors.at("Weight"_id) += adjustment.value();
+    }
+
+    // Get already computed values.
+    double H = 0; // TODO: person.risk_factors.at("Height"_id);
+    double BW = person.risk_factors.at("Weight"_id);
+    double PAL = person.risk_factors.at("PhysicalActivity"_id);
+    double EI = person.risk_factors.at("EnergyIntake"_id);
+
+    // Body fat.
+    double F;
+    if (person.gender == core::Gender::female) {
+        F = BW * (0.14 * person.age + 39.96 * log(BW / pow(H / 100, 2.0)) - 102.01) / 100.0;
+    } else if (person.gender == core::Gender::male) {
+        F = BW * (0.14 * person.age + 37.31 * log(BW / pow(H / 100, 2.0)) - 103.94) / 100.0;
+    } else {
+        throw core::HgpsException("Unknown gender");
+    }
 
-    // TODO: Compute physical activity level.
-    double PAL = PAL_0;
+    if (F < 0.0) {
+        // HACK: deal with negative body fat.
+        F = 0.2 * BW;
+    }
 
-    // TODO: some nutrient values need saving from previous year.
+    // Glycogen, water and extracellular fluid.
+    double G = 0.01 * BW;
+    double W = 2.7 * G;
+    double ECF = 0.7 * 0.235 * BW;
 
-    // Updated energy intake and carbohydrate intake.
-    // TODO: above, set EI_0 and CI_0 to risk_factors[*_previous].
-    double EI = person.get_risk_factor_value("EnergyIntake"_id);
-    double CI = person.get_risk_factor_value("Carbohydrate"_id);
+    // Lean tissue.
+    double L = BW - F - G - W - ECF;
 
-    // Compute glycogen and water.
-    double G = compute_G(CI, CI_0, G_0);
-    double W = compute_W(G);
+    // Model intercept value.
+    double delta = compute_delta(person.age, person.gender, PAL, BW, H);
+    double K = EI - (gamma_F * F + gamma_L * L + delta * BW);
 
-    // Compute extracellular fluid.
-    double ECF = compute_ECF(EI, EI_0, CI, CI_0, ECF_0);
+    // Compute energy expenditure.
+    double C = 10.4 * rho_L / rho_F;
+    double p = C / (C + F);
+    double x = p * eta_L / rho_L + (1.0 - p) * eta_F / rho_F;
+    double EE = compute_EE(BW, F, L, EI, K, delta, x);
+
+    // Set new state.
+    person.risk_factors["BodyFat"_id] = F;
+    person.risk_factors["LeanTissue"_id] = L;
+    person.risk_factors["Glycogen"_id] = G;
+    person.risk_factors["ExtracellularFluid"_id] = ECF;
+    person.risk_factors["Intercept_K"_id] = K;
+
+    // Set new energy expenditure (may not exist yet).
+    person.risk_factors["EnergyExpenditure"_id] = EE;
+}
+
+void KevinHallModel::kevin_hall_run(Person &person) const {
```

Comment:
Same here - not your fault! All of the formulas and calculations here are somewhat meaningless without an explanation, even dangerous because there's no way of checking their correctness.

Reply:
Agreed.

Reply:
In an ideal world, they would also help us write meaningful tests for them.